Terminal.Gui/ConsoleDrivers/WindowsDriver.cs

```diff
@@ -875,14 +877,150 @@ private void ChangeWin (Object s, SizeChangedEventArgs e)
+			return ConsoleKeyMapping.MapKeyModifiers (keyInfo, Key.PageDown);
+		case ConsoleKey.Enter:
+			return ConsoleKeyMapping.MapKeyModifiers (keyInfo, Key.Enter);
+		case ConsoleKey.Spacebar:
+			return ConsoleKeyMapping.MapKeyModifiers (keyInfo, keyInfo.KeyChar == 0 ? Key.Space : (Key)keyInfo.KeyChar);
+		case ConsoleKey.Backspace:
+			return ConsoleKeyMapping.MapKeyModifiers (keyInfo, Key.Backspace);
+		case ConsoleKey.Delete:
+			return ConsoleKeyMapping.MapKeyModifiers (keyInfo, Key.DeleteChar);
+		case ConsoleKey.Insert:
+			return ConsoleKeyMapping.MapKeyModifiers (keyInfo, Key.InsertChar);
+		case ConsoleKey.PrintScreen:
+			return ConsoleKeyMapping.MapKeyModifiers (keyInfo, Key.PrintScreen);
+
+		//case ConsoleKey.NumPad0:
+		//	return keyInfoEx.NumLock ? Key.D0 : Key.InsertChar;
+		//case ConsoleKey.NumPad1:
+		//	return keyInfoEx.NumLock ? Key.D1 : Key.End;
+		//case ConsoleKey.NumPad2:
+		//	return keyInfoEx.NumLock ? Key.D2 : Key.CursorDown;
+		//case ConsoleKey.NumPad3:
+		//	return keyInfoEx.NumLock ? Key.D3 : Key.PageDown;
+		//case ConsoleKey.NumPad4:
+		//	return keyInfoEx.NumLock ? Key.D4 : Key.CursorLeft;
+		//case ConsoleKey.NumPad5:
+		//	return keyInfoEx.NumLock ? Key.D5 : (Key)((uint)keyInfo.KeyChar);
+		//case ConsoleKey.NumPad6:
+		//	return keyInfoEx.NumLock ? Key.D6 : Key.CursorRight;
+		//case ConsoleKey.NumPad7:
+		//	return keyInfoEx.NumLock ? Key.D7 : Key.Home;
+		//case ConsoleKey.NumPad8:
+		//	return keyInfoEx.NumLock ? Key.D8 : Key.CursorUp;
+		//case ConsoleKey.NumPad9:
+		//	return keyInfoEx.NumLock ? Key.D9 : Key.PageUp;
+
+		case ConsoleKey.Oem1:
+		case ConsoleKey.Oem2:
+		case ConsoleKey.Oem3:
+		case ConsoleKey.Oem4:
+		case ConsoleKey.Oem5:
+		case ConsoleKey.Oem6:
+		case ConsoleKey.Oem7:
+		case ConsoleKey.Oem8:
+		case ConsoleKey.Oem102:
+		case ConsoleKey.OemPeriod:
+		case ConsoleKey.OemComma:
+		case ConsoleKey.OemPlus:
+		case ConsoleKey.OemMinus:
+			return (Key)((uint)keyInfo.KeyChar);
+		}
+
+		var key = keyInfo.Key;
+		var alphaBase = ((keyInfo.Modifiers == ConsoleModifiers.Shift) ^ (keyInfoEx.CapsLock)) ? 'A' : 'a';
+
+		if (key >= ConsoleKey.A && key <= ConsoleKey.Z) {
+			var delta = key - ConsoleKey.A;
+			if (keyInfo.Modifiers == ConsoleModifiers.Control) {
+				return (Key)(((uint)Key.CtrlMask) | ((uint)Key.A + delta));
+			}
+			if (keyInfo.Modifiers == ConsoleModifiers.Alt) {
+				return (Key)(((uint)Key.AltMask) | ((uint)Key.A + delta));
+			}
+			if (keyInfo.Modifiers == (ConsoleModifiers.Shift | ConsoleModifiers.Alt)) {
+				return ConsoleKeyMapping.MapKeyModifiers (keyInfo, (Key)((uint)Key.A + delta));
+			}
+			if ((keyInfo.Modifiers & (ConsoleModifiers.Alt | ConsoleModifiers.Control)) != 0) {
+				if (keyInfo.KeyChar == 0 || (keyInfo.KeyChar != 0 && keyInfo.KeyChar >= 1 && keyInfo.KeyChar <= 26)) {
+					return ConsoleKeyMapping.MapKeyModifiers (keyInfo, (Key)((uint)Key.A + delta));
+				}
+			}
+			return (Key)((uint)alphaBase + delta);
+			//return (Key)((uint)keyInfo.KeyChar);
```

Comment:
When a char can't be resolved by the delta and if `KeyChar` is different of zero, then the `KeyChar` must be used. Is the case where accent, É, @, €, etc...
So, this need to be uncommented and remove the `return (Key)((uint)alphaBase + delta);`

Reply:
When I'm trying to type `é` in the `Editor` scenario I get this error:
![image](https://github.com/gui-cs/Terminal.Gui/assets/13117724/b3c82ae1-5550-471c-bebb-8ff4f330e674)

So, what I'm trying to say is that characters with accent doesn't works on `WindowsDriver`, at least.
One thing is managing the `FakeDriver` to output whatever we want, but another thing is handling what a specific API is sending to the console.

Reply:
The cause of this error is because the `OnInvokeKeyBindings` is receiving the `Key.Null` on the `KeyEventArgs` and none of the shortcuts can have the `Key.Null`.
The cause of `OnInvokeKeyBindings` is being receiving a `Key.Null` is when the ´ accent is pressed. It can't be processed at this phase, normally it will receive another char, like `e` to result in the `é`.

Reply:
Before when `if (map == (Key)0xffffffff) {` only call the `OnKeyDown` when `inputEvent.KeyEvent.bKeyDown` and `OnKeyUp`, otherwise. Now `map` isn't checking this and all pass. Also the `OnKeyPressed` only was being called on the `inputEvent.KeyEvent.bKeyDown` and only the `OnKeyUp` was being called when the key is released and now the `OnKeyPressed` is called no matter is key down or not. This mustn't be the correct behavior, otherwise for e.g. pressing `´` + `e` must output `é` and not it's outputting `e` when the key is released.
The value of `(Key)0xffffffff` was used to return an invalid key which normally is used if Ctrl, Shift, Alt, etc. was pressed, but it can be also the `Key.Unknown`.

Reply:
> I broke ctrl-v too. Maybe you can explain what's going on. In any textview it double pastes.

The current behavior the `OnKeyPressed` is running twice and that why it doubles pastes.
This should be changed to `OnKeyPress`, like the event handler `KeyPress` and only when the key is down. When the key is released only the `OnKeyUp` must be called. This isn't like the mouse button where after a button press and then released will generate a button clicked.

Reply:
Also when I press `Shif+Ctrl+Alt` keys in the `Keys` scenario they appear separated, only they are joined when I press with another alpha character. If a user only want to check if `Shift+Ctrl` are pressed he can't.

Reply:
> Reference

I appreciate your help. 

In trying to simplify this I perhaps went too far. 

Rethinking things...

Reply:
> > I broke ctrl-v too. Maybe you can explain what's going on. In any textview it double pastes.
> 
> The current behavior the `OnKeyPressed` is running twice and that why it doubles pastes. This should be changed to `OnKeyPress`, like the event handler `KeyPress` and only when the key is down. When the key is released only the `OnKeyUp` must be called. This isn't like the mouse button where after a button press and then released will generate a button clicked.

First, i found the terminology in the old design inconsistent and confusing. I have attempted to be super-clear and consistent, and I think the current codebase IS clear and consistent. But based on what you wrote above, it may not be CORRECT (which is even more important!). 

With CursesDriver I understand it's not possible to see key-down (a key has been pressed) and key-up (a key that was previously pressed has been released) events. Only key-pressed (a key was pressed and then released).

While Windows does expose those, I wanted the design to be optimized for the most common use case, which is "a key was pressed and released". Thus my design says:

<img width="385" alt="image" src="https://github.com/gui-cs/Terminal.Gui/assets/585482/bbe8f59b-c9e7-458c-8841-4952bb8e594a">

Are you saying that there's a fourth concept: key-press (current tense, not past tense like key-pressED)? If so, how should we define it?

Reply:
> Are you saying that there's a fourth concept: key-press (current tense, not past tense like key-pressED)? If so, how should we define it?

No, that is how mouse sequence work and it's confuse the concept. The order is wrong, should be as before, `KeyDown`, `KeyPress` and `KeyUp`. `KeyDown` and `KeyPress` are called on key pressed and `KeyUp` is called on key released. If a user want to inspect if on a special key was pressed, `Shift`, `Ctrl`, `Alt`, `CapsLock` or `NumLock` he must use the `KeyDown` event. If a user want to inspect all the specials key and the alpha/numeric was pressed he must use the `KeyPress` event. If a user want to inspect all the keys released he must use the `KeyUp` event.
This confusion come a long time ago when a user open an issue saying that key managements should have a `KeyPressed` event but that's wrong. For that the `KeyUp` is used.